src/hotspot/share/gc/z/zRelocationSet.cpp

```diff
@@ -22,35 +22,102 @@
+  volatile size_t                _medium_next;
+
+  void install(ZForwarding* forwarding, volatile size_t* next) {
+    const size_t index = Atomic::fetch_and_add(next, 1u);
+    _forwardings[index] = forwarding;
+  }
+
+  void install_small(ZForwarding* forwarding) {
+    install(forwarding, &_small_next);
+  }
+
+  void install_medium(ZForwarding* forwarding) {
+    install(forwarding, &_medium_next);
+  }
+
+public:
+  ZRelocationSetInstallTask(ZForwardingAllocator* allocator, const ZRelocationSetSelector* selector) :
+      ZTask("ZRelocationSetInstallTask"),
+      _allocator(allocator),
+      _forwardings(NULL),
+      _nforwardings(selector->small()->length() + selector->medium()->length()),
+      _small_iter(selector->small()),
+      _medium_iter(selector->medium()),
+      _small_next(selector->medium()->length()),
+      _medium_next(0) {
+
+    // Reset the allocator to have room for the relocation
+    // set, all forwardings, and all forwarding entries.
+    const size_t relocation_set_size = _nforwardings * sizeof(ZForwarding*);
+    const size_t forwardings_size = _nforwardings * sizeof(ZForwarding);
+    const size_t forwarding_entries_size = selector->forwarding_entries() * sizeof(ZForwardingEntry);
+    _allocator->reset(relocation_set_size + forwardings_size + forwarding_entries_size);
 
-void ZRelocationSet::populate(ZPage* const* group0, size_t ngroup0,
-                              ZPage* const* group1, size_t ngroup1) {
-  _nforwardings = ngroup0 + ngroup1;
-  _forwardings = REALLOC_C_HEAP_ARRAY(ZForwarding*, _forwardings, _nforwardings, mtGC);
+    // Allocate relocation set
+    _forwardings = new (_allocator->alloc(relocation_set_size)) ZForwarding*[_nforwardings];
+  }
+
+  virtual void work() {
+    // Allocate and install forwardings for small pages
+    for (ZPage* page; _small_iter.next(&page);) {
+      ZForwarding* const forwarding = ZForwarding::alloc(_allocator, page);
+      install_small(forwarding);
+    }
 
-  size_t j = 0;
+    // Allocate and install forwardings for medium pages
+    for (ZPage* page; _medium_iter.next(&page);) {
+      ZForwarding* const forwarding = ZForwarding::alloc(_allocator, page);
+      install_medium(forwarding);
+    }
+  }
 
-  // Populate group 0
-  for (size_t i = 0; i < ngroup0; i++) {
-    _forwardings[j++] = ZForwarding::create(group0[i]);
+  ZForwarding** forwardings() const {
+    return _forwardings;
   }
 
-  // Populate group 1
-  for (size_t i = 0; i < ngroup1; i++) {
-    _forwardings[j++] = ZForwarding::create(group1[i]);
+  size_t nforwardings() const {
+    return _nforwardings;
   }
+};
+
+ZRelocationSet::ZRelocationSet(ZWorkers* workers) :
+    _workers(workers),
+    _allocator(),
+    _forwardings(NULL),
+    _nforwardings(0) {}
+
+void ZRelocationSet::install(const ZRelocationSetSelector* selector) {
+  // Install relocation set
+  ZRelocationSetInstallTask task(&_allocator, selector);
+  _workers->run_concurrent(&task);
+
```

Comment:
Now all space in `_allocator` is used; I wonder if it's worth placing an assert here, sth like `assert(_allocator.used() == _allocator.size())`.

Reply:
Sounds like a good idea. I'll add that to #804.